sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala

```diff
@@ -1847,6 +1847,36 @@ object GpuOverrides {
         override def convertToGpu(): Expression =
           sortOrder.withNewChildren(childExprs.map(_.convertToGpu()))
       }),
+    expr[PivotFirst](
+      "PivotFirst operator",
+      ExprChecks.groupByAgg(
+        TypeSig.commonCudfTypes + TypeSig.NULL + TypeSig.DECIMAL +
+          TypeSig.ARRAY.nested(TypeSig.commonCudfTypes + TypeSig.DECIMAL),
+        TypeSig.all,
+        Seq(ParamCheck(
+          "pivotColumn",
+          TypeSig.commonCudfTypes + TypeSig.NULL + TypeSig.DECIMAL,
+          TypeSig.all),
+          ParamCheck("valueColumn",
+          TypeSig.commonCudfTypes + TypeSig.NULL + TypeSig.DECIMAL,
+          TypeSig.all))),
+      (pivot, conf, p, r) => new ImperativeAggExprMeta[PivotFirst](pivot, conf, p, r) {
+        override def tagExprForGpu(): Unit = {
+          if (conf.hasNans &&
+            (pivot.pivotColumn.dataType.equals(FloatType) ||
+              pivot.pivotColumn.dataType.equals(DoubleType))) {
+            willNotWorkOnGpu("Pivot expressions over floating point columns " +
+              "that may contain NaN is disabled. You can bypass this by setting " +
+              s"${RapidsConf.HAS_NANS}=false")
+          }
+          // If pivotColumnValues doesn't have distinct values, follow back to CPU
+          if (pivot.pivotColumnValues.distinct.lengthCompare(pivot.pivotColumnValues.length) != 0) {
```

Comment:
Did this show up in testing? I don't see how this would ever happen as it would cause issues in the CPU code too.

Reply:
Yes, it showed up in testing. The result is null for all non-distinct entries in CPU.

```
spark.sql("""SELECT * FROM courseSales
PIVOT (
  sum(earnings)
  FOR year IN (2012, 2013,2012)
)""").show

+------+----+-----+----+                                                        
|course|2012| 2013|2012|
+------+----+-----+----+
|dotNET|null|48000|null|
|  Java|null|30000|null|
+------+----+-----+----+
```
In plugin, it throws cudfException:
```
  Caused by: ai.rapids.cudf.CudfException: vector::_M_range_check: __n (which is 3) >= this->size() (which is 3)
  at ai.rapids.cudf.Table.groupByAggregate(Native Method)
  at ai.rapids.cudf.Table.access$2900(Table.java:45)
  at ai.rapids.cudf.Table$AggregateOperation.aggregate(Table.java:2377)
  at com.nvidia.spark.rapids.GpuHashAggregateExec.computeAggregate(aggregate.scala:821)
  at com.nvidia.spark.rapids.GpuHashAggregateExec.$anonfun$doExecuteColumnar$1(aggregate.scala:431)
```

If we remove additional `2012` from the query, we get correct result for both CPU and GPU:
```
+------+-----+-----+
|course| 2012| 2013|
+------+-----+-----+
|dotNET|15000|48000|
|  Java|20000|30000|
+------+-----+-----+
```

Reply:
hmm, that is really interesting.  We could definitely match that same functionality, but why would anyone want that??? Falling back to the CPU feels like the right thing to do.